map_redcap_to_flywheel_service/utils/RCrecords_classes.py

```diff
@@ -0,0 +1,161 @@
+        from all the forms.  The first json object will only have values for the
+        non-repeating fields, the second object will only have values for the
+        first fillout of the repeating form, the third object will only have values
+        for the second fillout of the repeating form.
+
+        This checks through the records and looks for form names that are repeating.
+        Having these values makes it easier to handle repeating fields later.
+
+        For the example above it could return the 3 jsons:
+        {
+            'record_id': 'ID',
+            'redcap_repeat_instrument': '',
+            'redcap_repeat_instance': '',
+            'single_form_field': 'value'
+            'repeating_form_field': ''
+        },
+        {
+            'record_id': 'ID',
+            'redcap_repeat_instrument': 'repeating_form',
+            'redcap_repeat_instance': '1',
+            'single_form_field': ''
+            'repeating_form_field': 'first value'
+        },
+        {
+            'record_id': 'ID',
+            'redcap_repeat_instrument': 'repeating_form',
+            'redcap_repeat_instance': '2',
+            'single_form_field': ''
+            'repeating_form_field': 'second value'
+        }
+
+
+        """
+        repeating_forms = []
+        if len(self.records) == 1:
+            return repeating_forms
+
+        repeating_forms = set([r.get("redcap_repeat_instrument") for r in self.records if r .get("redcap_repeat_instrument")])
+
+        repeating_fields = set([field.get('field_name') for field in self.full_meta if field.get('form_name') in repeating_forms])
+        [repeating_fields.add(f"{form}_complete") for form in repeating_forms]
+
+        return repeating_fields
+
+
+    @classmethod
+    def GetRecordGroups(cls, full_meta, field_names, records, id_key):
+        """ Group records by identifying record ID.
+
+        Args:
+            full_meta (dict): the full field metadata from Project.metadata
+            field_names (list): extra metadata information on field names and field types
+            records (list): a list of redcap records, which are dicts
+            id_key (str): the field name in redcap that identifies the record "id".
+
+        Returns:
+            record_groups (dict): a dictionary in the format {redcap ID: record object for that ID}
+
+        """
+
+        # Get all the unique record IDs
+        uids = set([r.get(id_key) for r in records])
+        log.info(f"found {len(records)} records")
+        log.info(f"found {len(uids)} unique IDs")
+
+        record_groups = {}
+
+        # Loop through and extract matching records
+        for id in uids:
+            user_records = [r for r in records if r.get(id_key) == id]
+            # Create the record object for this group of records.
+            record_groups[id] = RedcapRecord(full_meta, field_names, user_records)
+
+        return record_groups
+
+
+    def get_record_dicts(self, fields_of_interest):
```

Comment:
This could be re-used as kind of a util tool to group the raw record jsons by ID's , if for some reason you ever needed to do that.

Basically though, the top one doesn't require "self", the others do.

Reply:
It's also kind of used as an initializer for itself, without overwriting the "__init__" function...I Think this just allows more flexibility but I'm open to other options!